File: v4/rust/iprange-livedb/src/range_store_cursor.rs
//! Ordered range traversal through a mutable mapped-page store.

use crate::contract::MetaV4;
use crate::error::{Error, Result};
use crate::fixed_tree::{self, CursorDirection, PageSource, Store};
use crate::key::IpKey;
use crate::range_cursor::{DirectRange, RangeItem};
use crate::range_tree::RangeCodec;

pub(crate) struct Cursor<K> {
    meta: MetaV4,
    release_private: bool,
    inner: fixed_tree::Cursor<RangeCodec<K>>,
}

impl<K: IpKey> Cursor<K> {
    pub(crate) fn new<S: Store>(
        store: &mut S,
        meta: &MetaV4,
        release_private: bool,
    ) -> Result<Self> {
        if meta.address_family != K::FAMILY {
            return Err(Error::WrongAddressFamily(
                "stored range cursor has the wrong address family",
            ));
        }
        if release_private
            && (meta.txn_id != store.target_txn() || meta.page_count != store.page_limit())
        {
            return Err(Error::Corrupt(
                "consumed range tree is outside the draft generation",
            ));
        }
        let inner = if release_private {
            fixed_tree::Cursor::new_consuming(store, meta.range_root, CursorDirection::Forward)?
        } else {
            fixed_tree::Cursor::new(
                &SelectedStore::new(&*store, *meta),
                meta.range_root,
                CursorDirection::Forward,
            )?
        };
        Ok(Self {
            meta: *meta,
            release_private,
            inner,
        })
    }

    pub(crate) fn next<S: Store>(&mut self, store: &mut S) -> Result<Option<DirectRange<K>>> {
        let next = if self.release_private {
            self.inner.next_consuming(store, &mut RangeItem)?
        } else {
            self.inner
                .next(&SelectedStore::new(&*store, self.meta), &mut RangeItem)?
        };
        crate::work::range_consumed(u64::from(next.is_some()));
        Ok(next)
    }
}

struct SelectedStore<'a, S> {
    store: &'a S,
    meta: MetaV4,
}

impl<'a, S> SelectedStore<'a, S> {
    const fn new(store: &'a S, meta: MetaV4) -> Self {
        Self { store, meta }
    }
}

impl<S: Store> PageSource for SelectedStore<'_, S> {
    type Page<'a>
        = S::ReadPage<'a>
    where
        Self: 'a;

    fn selected_txn(&self) -> u64 {
        self.meta.txn_id
    }

    fn selected_page_limit(&self) -> u64 {
        self.meta.page_count
    }

    fn view_page<'a, T, F>(&'a self, page_number: u32, inspect: F) -> Result<T>
    where
        F: FnOnce(Self::Page<'a>) -> Result<T>,
    {
        if page_number < 2 || u64::from(page_number) >= self.meta.page_count {
            return Err(Error::Corrupt("stored range page is outside its source"));
        }
        self.store.inspect_page(page_number, inspect)
    }
}

File: v4/go/internal/writer/range_store_cursor.go
// Ordered range traversal through a mutable mapped-page store (Rust
// range_store_cursor.rs): a forward cursor over one base generation read
// through the draft mapping. The source selection pins the base meta
// transaction and page count (Rust SelectedStore), so an ordered merge
// compares the incoming scan against the committed destination exactly,
// never against the draft's own pages. History never consumes the base
// tree (recorded chunk-3b-4 decision), so this cursor only reads, and
// next hands back one record by value so the merge loop allocates
// nothing per record (Rust Cursor::next returns DirectRange by value).

package writer

import (
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/tree"
	"github.com/firehol/iprange/v4/go/internal/work"
)

// selectedStore is one base generation pinned for reads through the
// draft mapping (Rust SelectedStore): every page view must name a page
// below the base page count, and pages parse against the base
// transaction, so a draft-grown file never leaks draft pages into a
// committed-generation scan.
type selectedStore struct {
	store *DraftStore
	meta  format.Meta
}

// TargetTxn returns the pinned base transaction (Rust
// PageSource::selected_txn).
func (s selectedStore) TargetTxn() uint64 { return s.meta.TxnID }

// PageLimit returns the pinned base page count (Rust
// PageSource::selected_page_limit).
func (s selectedStore) PageLimit() uint64 { return s.meta.PageCount }

// Inspect validates the page against the base bounds, then views it
// through the draft mapping (Rust SelectedStore::view_page).
func (s selectedStore) Inspect(pageNumber uint32) ([]byte, error) {
	if pageNumber < 2 || uint64(pageNumber) >= s.meta.PageCount {
		return nil, corrupt("stored range page is outside its source")
	}
	return s.store.InspectBase(pageNumber)
}

// DiscardPrivate is unreachable on a selected base source (the cursor is
// never consuming here); it fails closed if a future caller misuses it.
func (s selectedStore) DiscardPrivate(pageNumber uint32) error {
	return corrupt("selected base source cannot discard pages")
}

// The mutating Store methods are unreachable on a pinned generation
// source (base reads and catalog queries never allocate or update); each
// fails closed like DiscardPrivate, so a future misuse cannot corrupt a
// pinned generation through this view.
func (s selectedStore) Allocate() (uint32, error) {
	return 0, corrupt("selected base source cannot allocate")
}

func (s selectedStore) Update(pageNumber uint32) ([]byte, uint32, error) {
	return nil, 0, corrupt("selected base source cannot update")
}

func (s selectedStore) RestoreDirty(pageNumber uint32, tag uint32) error {
	return corrupt("selected base source cannot restore")
}

func (s selectedStore) CopyPage(source, destination uint32) ([]byte, []byte, uint32, error) {
	return nil, nil, 0, corrupt("selected base source cannot copy")
}

// rangeCursor is one forward cursor over a base generation range tree
// through the draft mapping (Rust Cursor<K> with SelectedStore).
type rangeCursor[K any] struct {
	cursor *tree.ForwardCursor[rangeRecord[K]]
	family rangeFamily[K]
}

// newRangeCursor opens the forward cursor over the source range tree
// (Rust Cursor::new). A non-consuming cursor reads the base generation
// through the selected store (history projection, ordered merges); a
// consuming cursor reads a draft-private tree directly and proves every
// page was born in the draft transaction (Rust release_private), which
// the coverage merge of the feed workflows uses.
func newRangeCursor[K any](store *DraftStore, base format.Meta, codec rangeFamily[K], consume bool) (*rangeCursor[K], error) {
	var source tree.ForwardStore
	if consume {
		if base.TxnID != store.draft.meta.TxnID || base.PageCount != store.draft.meta.PageCount {
			return nil, corrupt("consumed range tree is outside the draft generation")
		}
		source = store
	} else {
		source = selectedStore{store: store, meta: base}
	}
	cursor, err := tree.NewForwardCursor[rangeRecord[K]](codec, source, base.RangeRoot, consume)
	if err != nil {
		return nil, err
	}
	return &rangeCursor[K]{cursor: cursor, family: codec}, nil
}

// next returns the next range record in ascending key order, or ok=false
// at the end (Rust Cursor::next + RangeItem; one record by value).
func (c *rangeCursor[K]) next() (record rangeRecord[K], ok bool, err error) {
	record, ok, err = c.cursor.Next()
	if err != nil {
		return rangeRecord[K]{}, false, err
	}
	if !ok {
		return rangeRecord[K]{}, false, nil
	}
	work.RangeConsumed(1)
	return record, true, nil
}

// InspectBase views one mapped page without the base-bound validation
// (the DraftStore-bound readers use Inspect; the selected base source
// adds its own bound on top).
func (s *DraftStore) InspectBase(pageNumber uint32) ([]byte, error) {
	limit := s.committedPageCount
	if s.draft != nil {
		limit = s.draft.meta.PageCount
	}
	if uint64(pageNumber) >= limit {
		return nil, corrupt("draft page is out of bounds")
	}
	return s.mapping.Page(pageNumber)
}
